### PR TITLE
Update RCF to v3.8 and Enable Auto AD with 'Alert Once' Option

### PR DESCRIPTION
**Pull request by kaituo**
### Description
This PR added support for automatic Anomaly Detection (AD) and the 'Alert Once' option introduced in RCF 3.8.

Testing done:

1. Deserialization Test:
* Verified model deserialization from 3.0-rc3.
* Ensured consistent scoring using the rc3 checkpoint and rc3 dependency on identical data.

2. Backward Compatibility Test:
* Executed a mixed cluster with versions 2.10 and 3.0.
* Validated that older detectors still produce results without throwing any exceptions in a blue-green simulation scenario.

### Issues Resolved
https://github.com/opensearch-project/anomaly-detection/issues/978

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/ml/ModelManager.java [hunk 1]
```diff
@@ -532,6 +533,10 @@ private void trainModelForStep(
             .boundingBoxCacheFraction(TimeSeriesSettings.REAL_TIME_BOUNDING_BOX_CACHE_RATIO)
             .shingleSize(detector.getShingleSize())
             .anomalyRate(1 - thresholdMinPvalue)
+            .transformMethod(TransformMethod.NORMALIZE)
+            .alertOnce(true)
+            .autoAdjust(true)
```
Comment (sudiptoguha) [inline src/main/java/org/opensearch/ad/ml/ModelManager.java:538]: AlertOnce and AutoAdjust can interfere. It's ok to have both true -- will have fewer anomalies. If any of these are changed to false then we should consider  changing the other (in particular, if autoAdjust is false then do we want alertOnce -- unclear, may be may be not)

File: src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java [hunk 2]
```diff
@@ -256,6 +257,10 @@ private ThresholdedRandomCutForest createTRCF() {
             .precision(Precision.FLOAT_32)
             .randomSeed(seed)
             .boundingBoxCacheFraction(0)
+            .transformMethod(TransformMethod.NORMALIZE)
+            .alertOnce(true)
+            .autoAdjust(true)
+            .internalShinglingEnabled(false)
```
Comment (sudiptoguha) [inline src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java:263]: Is this for existing models? Single Streams? Creating external shingled models will take up more memory (and more queries)

**Review by sudiptoguha — APPROVED** [review]
Mostly ok - some questions which may need clarifications.

File: src/main/java/org/opensearch/ad/ml/ModelManager.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ml/ModelManager.java:538]: are you saying there are dependencies between these two options? How to figure out whether to set one false if the other one is false?

File: src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java (hunk 2 above)
Comment (kaituo) [reply src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java:263]: we only enabled internal shingling for real-time HC detectors. Other detectors all use external shingling for historical reasons.

Comment (sudiptoguha) [reply src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java:263]: Makes sense. Thx.

File: src/main/java/org/opensearch/ad/ml/ModelManager.java (hunk 1 above)
Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/ml/ModelManager.java:538]: Well, "alertOnce" is a directive :) That is, if there is a "run" or continuous anomalies then we start alert once and stop (till some internal checks indicates that at least one normal point was seen). In the autoAdjust case, we would learn and try to suppress anomalies of both the observed value and expected value are in some noise range. So if there is a level shift followed quickly by another spike/dip then autoAdjust has some chance of catching it. But all of these are based on unsupervised estimations -- so it's hard to guarantee behavior. In general starting from (false,false) autoAdjust = true should reduce anomalies; and alertOnce = true would reduce even further. But there may be applications where users prefer a lot of continuous anomalies (for example, to detect a level shift!).

File: src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java [hunk 3]
```diff
@@ -1098,4 +1100,138 @@ public void testShouldSave() {
...
+        URI uri = ClassLoader.getSystemResource("org/opensearch/ad/ml/rcf_3_0_rc3_single_stream.json").toURI();
+        String filePath = Paths.get(uri).toString();
+        String json = Files.readString(Paths.get(filePath), Charset.defaultCharset());
+        // For the parsing of .toTrcf to work I had to manually change "\u003d" in code back to =.
+        // In the byte array it doesn't seem like this is an issue but whenever reading the byte array response into a file it
+        // converts "=" to "\u003d" https://groups.google.com/g/google-gson/c/JDHUo9DWyyM?pli=1
+        // I also needed to bypass the trcf as it wasn't being read as a key value but instead part of the string
+        Map map = gson.fromJson(json, Map.class);
+        String model = (String) ((Map) ((Map) ((ArrayList) ((Map) map.get("hits")).get("hits")).get(0)).get("_source")).get("modelV2");
+        // model = model.split(":")[1].substring(1);
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java:1120]: nit: previously i think I added this to deal with "trcf" was this not needed any more in this case?

File: src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java [hunk 4]
```diff
@@ -1098,4 +1100,138 @@ public void testShouldSave() {
...
+        // In the byte array it doesn't seem like this is an issue but whenever reading the byte array response into a file it
+        // converts "=" to "\u003d" https://groups.google.com/g/google-gson/c/JDHUo9DWyyM?pli=1
+        // I also needed to bypass the trcf as it wasn't being read as a key value but instead part of the string
+        Map map = gson.fromJson(json, Map.class);
+        String model = (String) ((Map) ((Map) ((ArrayList) ((Map) map.get("hits")).get("hits")).get(0)).get("_source")).get("modelV2");
+        model = model.split(":")[1];
+        model = model.substring(1, model.length() - 2);
+        // model = Base64.getEncoder().encodeToString(org.apache.commons.codec.binary.Base64.decodeBase64(model));
+        // Simulate JSON parsing by replacing Unicode escape sequence with the actual character
+        model = unescapeJavaString(model);
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java:1175]: Maybe related to how I record those Json response file. Without doing it, we experience model corruption exception.

[the remark this answers is not in the document]
Comment (amitgalitz) [reply src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java:1176]: I said the difference in recording the json response makes sense, I'm not 100% sure if we are talking about the same thing, but I used the lines above to remove the last = charachter in the response

**Review by amitgalitz — APPROVED** [review]

File: src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaituo) [reply src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java:1176]: yeah, different checkpoint might need different conversion. In my single stream checkpoint, I don't need to do anything.

**Review by sudiptoguha — APPROVED** [review]

**Pull request merged**